### PR TITLE
Add support for server-side encryption with customer key to S3

### PR DESCRIPTION
**Pull request by wendigo**
## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystem.java [hunk 1]
```diff
@@ -273,4 +274,71 @@ static boolean isUnrecoverableException(Throwable throwable)
         return Throwables.getCausalChain(throwable).stream()
                 .anyMatch(t -> t instanceof TrinoFileSystemException || t instanceof FileNotFoundException || t instanceof UnsupportedOperationException);
     }
+
+    /**
+     * Creates an encryped TrinoInputFile which can be used to read the encrypted file data.
```
Comment (electrum) [inline lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystem.java:279]: Typo "encryped"

File: lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystem.java [hunk 2]
```diff
@@ -273,4 +274,71 @@ static boolean isUnrecoverableException(Throwable throwable)
...
+     * The file location path cannot be empty, and must not end with a slash or whitespace.
+     *
+     * @throws IllegalArgumentException if location is not valid for this file system
+     * @throws UnsupportedOperationException if encrypted files are not supported
+     */
+    default TrinoInputFile newEncryptedInputFile(Location location, EncryptionKey key)
```
Comment (electrum) [inline lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystem.java:285]: Move these methods above, to be after the other `newInputFile` methods

File: lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystem.java [hunk 3]
```diff
@@ -273,4 +274,71 @@ static boolean isUnrecoverableException(Throwable throwable)
...
+    {
+        throw new UnsupportedOperationException("Encrypted files are not supported");
+    }
+
+    /**
+     * Creates a TrinoInputFile with a predeclared length which can be used to read the file data.
```
Comment (electrum) [inline lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystem.java:291]: This doesn't mention encryption but the one above does

File: lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystem.java [hunk 4]
```diff
@@ -273,4 +274,71 @@ static boolean isUnrecoverableException(Throwable throwable)
...
+     *
+     * @throws IllegalArgumentException if location is not valid for this file system
+     */
+    default TrinoInputFile newEncryptedInputFile(Location location, long length, EncryptionKey key)
+    {
+        throw new UnsupportedOperationException("Encrypted files are not supported");
+    }
+
+    /**
+     * Creates a TrinoInputFile with a predeclared length and lastModifiedTime which can be used to read the file data.
```
Comment (electrum) [inline lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystem.java:304]: Same here

File: lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystem.java [hunk 5]
```diff
@@ -273,4 +274,71 @@ static boolean isUnrecoverableException(Throwable throwable)
...
+    }
+
+    /**
+     * Creates an encrypted TrinoOutputFile which can be used to create or overwrite the file.
+     * The file location path cannot be empty, and must not end with a slash or whitespace.
+     *
+     * @throws IllegalArgumentException if location is not valid for this file system
+     * @throws UnsupportedOperationException if encrypted files are not supported
+     */
+    default TrinoOutputFile newEncryptedOutputFile(Location location, EncryptionKey key)
```
Comment (electrum) [inline lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystem.java:324]: Move above, after `newOutputFile`

File: lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystem.java [hunk 6]
```diff
@@ -273,4 +274,71 @@ static boolean isUnrecoverableException(Throwable throwable)
...
+     * Returns the direct encrypted pre-signed URI location for the given storage location.
+     * <p></p>
+     * Pre-signed URIs allow for retrieval of the files directly from the storage location.
+     * This is useful for large files where the server would be a bottleneck.
+     *
+     * @throws UnsupportedOperationException if the pre-signed URIs are not supported
+     * @return the pre-signed URI to the storage location or `Optional.empty()`
+     *         if pre-signed URI cannot be generated.
+     */
+    default Optional<UriLocation> encryptedPreSignedUri(Location location, Duration ttl, EncryptionKey key)
```
Comment (electrum) [inline lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystem.java:339]: Move above, after `preSignedUri`

File: lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystem.java [hunk 7]
```diff
@@ -273,4 +274,71 @@ static boolean isUnrecoverableException(Throwable throwable)
...
+     * @throws UnsupportedOperationException if encrypted files are not supported
+     */
+    default TrinoOutputFile newEncryptedOutputFile(Location location, EncryptionKey key)
+    {
+        throw new UnsupportedOperationException("Encrypted files are not supported");
+    }
+
+    /**
+     * Returns the direct encrypted pre-signed URI location for the given storage location.
+     * <p></p>
```
Comment (electrum) [inline lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystem.java:331]: Javadoc should just have `<p>`

File: lib/trino-filesystem/src/main/java/io/trino/filesystem/encryption/EncryptionEnforcingFileSystem.java [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    @Override
+    public TrinoInputFile newInputFile(Location location)
+    {
+        return newEncryptedInputFile(location, key);
+    }
+
+    @Override
+    public TrinoInputFile newEncryptedInputFile(Location location, EncryptionKey key)
```
Comment (electrum) [inline lib/trino-filesystem/src/main/java/io/trino/filesystem/encryption/EncryptionEnforcingFileSystem.java:54]: It seems like these methods should either

* always fail, as we expect the non-encryption methods to be called
* fail if the passed key doesn't match the instance key

File: lib/trino-filesystem/src/test/java/io/trino/filesystem/AbstractTestTrinoFileSystem.java [hunk 9]
```diff
@@ -141,19 +158,19 @@ void testInputFileMetadata()
             throws IOException
     {
         // an input file cannot be created at the root of the file system
-        assertThatThrownBy(() -> getFileSystem().newInputFile(getRootLocation()))
+        assertThatThrownBy(() -> getTestingFileSystem().newInputFile(getRootLocation()))
```
Comment (electrum) [inline lib/trino-filesystem/src/test/java/io/trino/filesystem/AbstractTestTrinoFileSystem.java:161]: Having both `getFileSystem()` and `getTestingFileSystem()` is confusing. It will be easy to use the wrong one in the test code. Let's find a better way.

I think we could do the wrapping in the implementations, which would also avoid all the changes in `AbstractTestS3FileSystem`.

File: lib/trino-filesystem-s3/pom.xml [hunk 10]
```diff
@@ -304,6 +304,7 @@
                         <configuration>
                             <excludes>
                                 <exclude>**/TestS3FileSystemAwsS3.java</exclude>
+                                <exclude>**/TestS3FileSystemAwsS3WithEncryption.java</exclude>
```
Comment (electrum) [inline lib/trino-filesystem-s3/pom.xml:307]: Or change the above to
```xml
<exclude>**/TestS3FileSystemAwsS3*.java</exclude>

File: lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystem.java [hunk 11]
```diff
@@ -322,6 +349,41 @@ public Optional<UriLocation> preSignedUri(Location location, Duration ttl)
         }
     }
 
+    @Override
+    public Optional<UriLocation> encryptedPreSignedUri(Location location, Duration ttl, EncryptionKey key)
```
Comment (electrum) [inline lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystem.java:353]: Make the above method delegate to this one, as it seems the only difference is setting the key

File: lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3InputFile.java [hunk 12]
```diff
@@ -105,6 +112,13 @@ private GetObjectRequest newGetObjectRequest()
                 .requestPayer(requestPayer)
                 .bucket(location.bucket())
                 .key(location.key())
+                .applyMutation(builder -> {
+                    if (key.isPresent()) {
+                        builder.sseCustomerKey(encoded(key.get()));
+                        builder.sseCustomerAlgorithm("AES256");
```
Comment (electrum) [inline lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3InputFile.java:118]: Shouldn't this be `key.getAlgorithm()`? Or it should verify the key algorithm?

File: lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3InputFile.java [hunk 13]
```diff
@@ -116,6 +130,13 @@ private boolean headObject()
                 .requestPayer(requestPayer)
                 .bucket(location.bucket())
                 .key(location.key())
+                .applyMutation(builder -> {
+                    if (key.isPresent()) {
+                        builder.sseCustomerKey(encoded(key.get()));
+                        builder.sseCustomerAlgorithm(AES256.toString());
```
Comment (electrum) [inline lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3InputFile.java:136]: Same here. Also, use the string or enum consistently.

File: lib/trino-filesystem/src/main/java/io/trino/filesystem/encryption/EncryptionKey.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+import static java.util.Objects.requireNonNull;
+
+public class EncryptionKey
+{
+    private final byte[] key;
+    private final String algorithm;
+
+    public EncryptionKey(byte[] key, String algorithm)
+    {
+        this.key = key;
```
Comment (electrum) [inline lib/trino-filesystem/src/main/java/io/trino/filesystem/encryption/EncryptionKey.java:27]: Clone as arrays are immutable

File: lib/trino-filesystem/src/main/java/io/trino/filesystem/encryption/EncryptionKey.java [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+        this.algorithm = requireNonNull(algorithm, "algorithm is null");
+    }
+
+    public byte[] getKey()
+    {
+        return key;
```
Comment (electrum) [inline lib/trino-filesystem/src/main/java/io/trino/filesystem/encryption/EncryptionKey.java:33]: Return a clone, so that callers can't modify the original

File: lib/trino-filesystem/src/main/java/io/trino/filesystem/encryption/EncryptionKey.java [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.filesystem.encryption;
+
+import java.util.concurrent.ThreadLocalRandom;
+
+import static java.util.Objects.requireNonNull;
+
+public class EncryptionKey
```
Comment (electrum) [inline lib/trino-filesystem/src/main/java/io/trino/filesystem/encryption/EncryptionKey.java:20]: Why not a record? At least make it final

File: lib/trino-filesystem/src/main/java/io/trino/filesystem/encryption/EncryptionEnforcingFileSystem.java (hunk 8 above)
Comment (wendigo) [reply lib/trino-filesystem/src/main/java/io/trino/filesystem/encryption/EncryptionEnforcingFileSystem.java:54]: Added a check

**Comment by wendigo** [comment]
AC @electrum

File: lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3SseCUtils.java [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+
+    public static String encoded(EncryptionKey key)
+    {
+        return Base64.getEncoder().encodeToString(key.key());
+    }
+
+    public static String md5Checksum(EncryptionKey key)
+    {
+        try {
+            MessageDigest digest = MessageDigest.getInstance("MD5");
```
Comment (electrum) [inline lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3SseCUtils.java:34]: Could use Guava `Hashing.md5()`

File: lib/trino-filesystem-s3/src/test/java/io/trino/filesystem/s3/AbstractTestS3FileSystem.java [hunk 18]
```diff
@@ -134,19 +146,30 @@ void testFileWithTrailingWhitespaceAgainstNativeClient()
...
+                            builder.sseCustomerKeyMD5(md5Checksum(randomEncryptionKey));
+                        }
+                    })
+                    .build();
+
+            s3Client.putObject(putObjectRequest, RequestBody.fromBytes(contents.clone()));
             try {
                 // Verify listing
-                List<FileEntry> listing = toList(fileSystem.listFiles(getRootLocation().appendPath("foo")));
+                List<FileEntry> listing = toList(getFileSystem().listFiles(getRootLocation().appendPath("foo")));
```
Comment (electrum) [inline lib/trino-filesystem-s3/src/test/java/io/trino/filesystem/s3/AbstractTestS3FileSystem.java:165]: Move this refactoring to a separate commit

File: lib/trino-filesystem/src/main/java/io/trino/filesystem/encryption/EncryptionEnforcingFileSystem.java [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+    {
+        verify(this.key.equals(key), "Provided key is not the same as the class encryption key");
```
Comment (electrum) [inline lib/trino-filesystem/src/main/java/io/trino/filesystem/encryption/EncryptionEnforcingFileSystem.java:57]: `checkArgument()` for all of these

File: lib/trino-filesystem/src/main/java/io/trino/filesystem/encryption/EncryptionKey.java [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+package io.trino.filesystem.encryption;
+
+import java.util.Arrays;
+import java.util.concurrent.ThreadLocalRandom;
+
+import static java.util.Objects.requireNonNull;
+
+public record EncryptionKey(byte[] key, String algorithm)
+{
+    public EncryptionKey {
```
Comment (electrum) [inline lib/trino-filesystem/src/main/java/io/trino/filesystem/encryption/EncryptionKey.java:23]: Formatting

File: lib/trino-filesystem/src/main/java/io/trino/filesystem/encryption/EncryptionKey.java [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+        requireNonNull(algorithm, "algorithm is null");
+        key = Arrays.copyOf(key, key.length);
```
Comment (electrum) [inline lib/trino-filesystem/src/main/java/io/trino/filesystem/encryption/EncryptionKey.java:25]:
```java
key = key.clone()

Comment (electrum) [reply lib/trino-filesystem/src/main/java/io/trino/filesystem/encryption/EncryptionKey.java:25]: Actually... let's just let it be mutable and only check for null here

Otherwise we have to copy it every time it is accessed

**Review by electrum — APPROVED** [review]

File: lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3SseCUtils.java (hunk 17 above)
Comment (wendigo) [reply lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3SseCUtils.java:34]: @electrum it is deprecated and generates a warning as error

File: lib/trino-filesystem-s3/src/test/java/io/trino/filesystem/s3/AbstractTestS3FileSystem.java (hunk 18 above)
Comment (wendigo) [reply lib/trino-filesystem-s3/src/test/java/io/trino/filesystem/s3/AbstractTestS3FileSystem.java:165]: Extracted

File: lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3SseCUtils.java [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+ */
+package io.trino.filesystem.s3;
+
+import io.trino.filesystem.encryption.EncryptionKey;
+
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Base64;
+
+final class S3SseCUtils
```
Comment (findinpath) [inline lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3SseCUtils.java:22]: I know what Sse means - Server side encryption

What does `C` stand for? Would it make sense to have the actual word  used instead?

File: lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystem.java [hunk 23]
```diff
@@ -63,6 +64,18 @@ public interface TrinoFileSystem
...
     TrinoInputFile newInputFile(Location location);
 
+    /**
+     * Creates an encrypted TrinoInputFile which can be used to read the encrypted file data.
+     * The file location path cannot be empty, and must not end with a slash or whitespace.
+     *
+     * @throws IllegalArgumentException if location is not valid for this file system
+     * @throws UnsupportedOperationException if server side encryption is not supported
+     */
+    default TrinoInputFile newEncryptedInputFile(Location location, EncryptionKey key)
```
Comment (findinpath) [inline lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystem.java:74]: The current PR is missing a broader context on where is this building block going to be used.

This could help a potential reviewer to connect the dots easier on why do we need to extend the `TrinoFileSystem` interface.

File: lib/trino-filesystem-s3/src/test/java/io/trino/filesystem/s3/AbstractTestS3FileSystem.java [hunk 24]
```diff
@@ -134,19 +146,30 @@ void testFileWithTrailingWhitespaceAgainstNativeClient()
         try (S3Client s3Client = createS3Client()) {
             String key = "foo/bar with whitespace ";
             byte[] contents = "abc foo bar".getBytes(UTF_8);
-            s3Client.putObject(
-                    request -> request.bucket(bucket()).key(key),
-                    RequestBody.fromBytes(contents.clone()));
+
+            PutObjectRequest putObjectRequest = PutObjectRequest.builder()
```
Comment (findinpath) [inline lib/trino-filesystem-s3/src/test/java/io/trino/filesystem/s3/AbstractTestS3FileSystem.java:150]: Curiosity question: Can we have for new table formats (Iceberg, Delta Lake) situations where the metadata files are not encrypted, but the content files would be encrypted?

Probably not because the metadata files contain stats which effectively contain content which should be encrypted. I just want to make sure that have this unlikely scenario on the radar.

File: lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3SseCUtils.java (hunk 22 above)
Comment (wendigo) [reply lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3SseCUtils.java:22]: Client

Comment (wendigo) [reply lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3SseCUtils.java:22]: Server Side Encryption with Client key (SSE-C). That's a common name.

File: lib/trino-filesystem-s3/src/test/java/io/trino/filesystem/s3/AbstractTestS3FileSystem.java (hunk 24 above)
Comment (wendigo) [reply lib/trino-filesystem-s3/src/test/java/io/trino/filesystem/s3/AbstractTestS3FileSystem.java:150]: This is outside of the scope of this PR, I think that parquet modular encryption has this scheme.

File: lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystem.java (hunk 23 above)
Comment (wendigo) [reply lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystem.java:74]: it will be used in the spooling. I've discussed API with @electrum already

**Comment by mosabua** [comment]
No RN @wendigo .. right?

**Comment by wendigo** [comment]
@mosabua No, this is used only for spooling

**Comment by mosabua** [comment]
Thanks @wendigo
